// sim-sim-js/lib/server_message_factory.js
(function() {
  var ServerMessageFactory;

  ServerMessageFactory = (function() {
    function ServerMessageFactory() {}

    ServerMessageFactory.prototype.event = function(id, data) {
      return {
        type: 'ServerMessage::Event',
        sourcePlayerId: id,
        data: data
      };
    };

    ServerMessageFactory.prototype.playerJoined = function(id) {
      return {
        type: 'ServerMessage::PlayerJoined',
        playerId: id
      };
    };

    ServerMessageFactory.prototype.playerLeft = function(id) {
      return {
        type: 'ServerMessage::PlayerLeft',
        playerId: id
      };
    };

    ServerMessageFactory.prototype.turnComplete = function(currentTurn) {
      return {
        type: 'ServerMessage::TurnComplete',
        turnNumber: currentTurn
      };
    };

    ServerMessageFactory.prototype.idAssigned = function(id) {
      return {
        type: 'ServerMessage::IdAssigned',
        ourId: id
      };
    };

    ServerMessageFactory.prototype.gamestateRequest = function(id) {
      return {
        type: 'ServerMessage::GamestateRequest',
        forPlayerId: id
      };
    };

    ServerMessageFactory.prototype.startGame = function(forPlayerId, turnPeriod, currentTurn, protoTurn, simState, worldState) {
      return {
        type: 'ServerMessage::StartGame',
        yourId: forPlayerId,
        turnPeriod: turnPeriod,
        currentTurn: currentTurn,
        protoTurn: protoTurn,
        simState: simState,
        worldState: worldState
      };
    };

    return ServerMessageFactory;

  })();

  module.exports = ServerMessageFactory;

}).call(this);
